### ISSUE TITLE
Disconnect external repositories

### ISSUE THREAD

**Issue by ghost** [webplates/nofw#5]
The current setup forces the user to utilize a certain dependency injector, routing library, etc. These are external libraries with no internal QA and may not suit everyone. Fastroute, for example, has no reverse routing functionality. Therefore the external libraries should be disconnected and the core should leave the user free to utilize the functionality any way they like, with a recommended, but not compulsory default setup.

This can be achieved by defining interfaces and objects for communication for each functionality provided by external libraries. A separate package can then be compiled that pulls in the recommended external libraries, but leaving the user free to chose their own setup if they are not satisfied with the defaults.

For example, in a similar work (Piccolo) this is the definition for a dependency injector: https://github.com/opsbears/piccolo/blob/master/src/DependencyInjection/DependencyInjectionContainer.php The core framework then depends on a metapackage named `piccolo-dic`, which is then implemented by `piccolo-dic-auryn`: https://github.com/opsbears/piccolo-dic-auryn/blob/master/src/AurynDependencyInjectionContainer.php

A similar approach  can be used for templating.

**Comment by tgabi333** [comment]
Thats why PHP-FIG exists. eg: psr-11 for DI

**Comment by ghost** [comment]
PHP-FIG 11 does not define a full dependency injector, just the configuration for it, so it does not enable this project to completely disconnect the injection library. On top of that, even though the documentation states otherwise, this interface they define highly encourages using it as a service locator, an antipattern that has been plaguing even well known frameworks like Symfony for years.

**Comment by sagikazarmark** [comment]
I had that in mind when I started the project, but I decided not to introduce such abstraction layers.

The main reason is that not all of these components are meant to be interoperable. Sticking to your example (routing): once you replace the router with another, it's not the same application anymore. The same is true for DI and templating as well.

The other reason is that introducing such layer would increase the configuration overhead which I would like to avoid. No abstract configuration format which needs to be documented. If you use FastRoute, you configure it in whatever way you choose, no restrictions.

The nofw skeleton is an opinionated example how such application can look like. It's not that much code, so the idea can easily be adopted to any other router/DI/templating solution. Again, avoiding extra configuration layers as much as possible is one of the main goals.

In fact, I would like to write documentation (recipes if you wish) about how you can replace one component with another or how you can add further functionality (like session storage or error monitoring).

The project is still in an early phase though, so this might change. For example, I started to work on an error handler abstraction which decouples the error monitoring from the application:

https://github.com/nofw/error

**Comment by ghost** [comment]
I found that things like templating, DIC, etc are easy to abstract and additional functionality can easily added by implementing additional interfaces. I would recommend considering this for at least routing, because it is sufficiently abstract and disconnected from the main app logic, and Fastroute, while really fast, is lacking in the feature department.

**Comment by sagikazarmark** [comment]
> I would recommend considering this for at least routing, because it is sufficiently abstract

I agree that in case of routing it might make sense to introduce that abstraction since the configuration itself is simple enough to be applied to any routing solution.

> I found that things like templating, DIC, etc are easy to abstract

It may not be hard to add an abstraction layer for these either, but:

- If you change the template engine, you have to change the templates themselves too. If you do that, the abstraction itself really does not have that much benefit.
- Since PSR-11 DiCs mostly differ in their configuration. Unfortunately they are quite configuration heavy, so even if they have the same features, it would be hard to provide them ALL under a single (opinionated) abstraction layer. Things might change if the FIG introduces a standard for that which gets implemented directly by the DiCs themselves.

I will think about the routing.

**Comment by sagikazarmark** [comment]
Another idea of mine is to have an installer which lets the user choosing the desired components during the project setup. Although this would not make replacing components easier, but would not force users to use FastRoute either.

**Comment by ghost** [comment]
I also found that injection configuration via the dependency injection layer makes sense. Each module can read the configuration (from environment or config file) and configure the classes it is responsible for. This makes it especially easy when classes request configuration in their constructor.

If you do it like this, each module can have its very own configuration structure and still be injectable and usable behind an interface.

**Comment by sagikazarmark** [comment]
I don't see a contradiction here. All you say is possible when the application relies on a specific DiC and it's configuration format. The only difference is that's not abstracted away.

What is NOT possible this way is to ship reusable modules between applications (at least not with 100% confidence), but that's a completely different story.

I understand that a user may or may not wish to use a specific component and I agree that it should be a choice of the user. But once that choice is made (probably after careful investigation and planning) I don't think that such core part of the application should be replaced with another.

I also understand that an abstraction would make that possible, but an abstraction is also a huge responsibility when it comes to the API design. It may not be able to cover all features of an implementation that a user may want to use, hence relying on its API can be a better choice.

Even if there are no such features, replacing the underlying implementation can still be a huge risk. Such "abstraction" cannot ensure that two implementations can perfectly substitute each other, since it only covers the configuration part.

And if it becomes absolutely necessary to replace a DiC implementation (abandoned, security risks, etc), you can use the fact that well-crafted containers (are going to) use PSR-11 today, so they *should* behave the same way which lets you do the transition as a careful, iterative process.

What other use cases are there for such abstraction other than allowing to replace a component with another?

Last, but not least: there is the argument of simplicity. Every extra layer makes the application more complex.

**Comment by ghost** [comment]
I think a distinction has to be made between things that a user most likely wants to replace (routing library, templating system) and things that most likely don't need replacing (DIC). For routing and templating it is fairly easy to define an interface to make the communication layer compatible.

For views, for example, I chose a simple hashmap (array with string keys) format to pass data to the view layer, and the file to use as a template is derived from the controller/action name. It is really simple to implement and leaves the user a choice to use whatever templating system they prefer without having to touch the controllers themselves. It also encourages best practices in terms of separation between view and controller.

The same goes for routing. The only job of a routing layer is to take a PSR-7 request and supply a class and method name or a callable that needs to be called as a result. Optionally, it can also supply a hashmap / array of extra data that has been generated from URL parameters. It does not need to take care of dispatching the call to the controller, and it also does not have to have a unified configuration. After all, that's why people use a different routing layer because different routing layers serve different purposes.

Ultimately the choice is yours, but from my own experience I would recommend doing a strict separation of external libraries wherever possible, and defining interfaces for standardized functionality. It will also help testability of internal components.

**Comment by sagikazarmark** [comment]
I still don't see the reason why a user would want to replace routing or templating engine once it's already settled.

Also, in case of routing a common configuration syntax might make this easy, but in case of templating you have to rewrite all templates as well. You can't abstract that.

I think it would make sense to have some kind of output abstraction passed to controllers so they don't have to know about the implementation, but I would rather see that as a separate library with standardization as the intent, not built into the skeleton.

> It does not need to take care of dispatching the call to the controller, and it also does not have to have a unified configuration.

I am a little bit confused here: I thought you suggested standard configuration so far.

Anyway, currently the routing layer is nothing more but a middleware in the PSR-15 middleware chain. If one would like to replace routing, it's only a middleware that needs to be replaced, given you don't expect the configuration format to be a standard one. What do you think?

**Comment by ghost** [comment]
You wouldn't wanna replace it once your project is up and running, you'd want to replace it when you start. For example because your team has vast experience in something else or you want something faster than Twig.

Regarding the configuration I suggested having a standardized place where you put the configuration, but the structure can be different depending on the library. In fact, it has to be.
